File: camerax_android/lib/src/core/dynamic_range_impl.dart
import 'package:camerax_android/src/api.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

final class DynamicRangeImpl implements DynamicRange {
  final DynamicRangeProxyApi api;

  DynamicRangeImpl.internal(this.api);

  @override
  Future<DynamicRange$BitDepth> getBitDepth() =>
      api.getBitDepth().then((e) => e.impl);

  @override
  Future<DynamicRange$Encoding> getEncoding() =>
      api.getEncoding().then((e) => e.impl);
}

final class DynamicRangeChannelImpl extends DynamicRangeChannel {
  @override
  DynamicRange get unspecifid =>
      DynamicRangeImpl.internal(DynamicRangeProxyApi.unspecifid);
  @override
  DynamicRange get sdr => DynamicRangeImpl.internal(DynamicRangeProxyApi.sdr);
  @override
  DynamicRange get hdrUnspecified10Bit =>
      DynamicRangeImpl.internal(DynamicRangeProxyApi.hdrUnspecified10Bit);
  @override
  DynamicRange get hdr10_10Bit =>
      DynamicRangeImpl.internal(DynamicRangeProxyApi.hdr10_10Bit);
  @override
  DynamicRange get hdr10Plus10Bit =>
      DynamicRangeImpl.internal(DynamicRangeProxyApi.hdr10Plus10Bit);
  @override
  DynamicRange get hlg10Bit =>
      DynamicRangeImpl.internal(DynamicRangeProxyApi.hlg10Bit);
  @override
  DynamicRange get dolbyVision8Bit =>
      DynamicRangeImpl.internal(DynamicRangeProxyApi.dolbyVision8Bit);
  @override
  DynamicRange get dolbyVision10Bit =>
      DynamicRangeImpl.internal(DynamicRangeProxyApi.dolbyVision10Bit);

  @override
  DynamicRange create({
    required DynamicRange$Encoding encoding,
    required DynamicRange$BitDepth bitDepth,
  }) {
    final api = DynamicRangeProxyApi(
      encoding: encoding.api,
      bitDepth: bitDepth.api,
    );
    return DynamicRangeImpl.internal(api);
  }
}

extension DynamicRange$EncodingX on DynamicRange$Encoding {
  DynamicRangeEncodingApi get api => DynamicRangeEncodingApi.values[index];
}

extension DynamicRangeEncodingApiX on DynamicRangeEncodingApi {
  DynamicRange$Encoding get impl => DynamicRange$Encoding.values[index];
}

extension DynamicRange$BitDepthX on DynamicRange$BitDepth {
  DynamicRangeBitDepthApi get api => DynamicRangeBitDepthApi.values[index];
}

extension DynamicRangeBitDepthApiX on DynamicRangeBitDepthApi {
  DynamicRange$BitDepth get impl => DynamicRange$BitDepth.values[index];
}

extension DynamicRangeX on DynamicRange {
  DynamicRangeProxyApi get api {
    final impl = this;
    if (impl is! DynamicRangeImpl) throw TypeError();
    return impl.api;
  }
}

extension DynamicRangeProxyApiX on DynamicRangeProxyApi {
  DynamicRange get impl => DynamicRangeImpl.internal(this);
}

File: camerax_android/android/src/main/kotlin/dev/zeekr/camerax_android/core/DynamicRangeImpl.kt
package dev.zeekr.camerax_android.core

import androidx.camera.core.DynamicRange
import dev.zeekr.camerax_android.CameraXApiPigeonProxyApiRegistrar
import dev.zeekr.camerax_android.DynamicRangeBitDepthApi
import dev.zeekr.camerax_android.DynamicRangeEncodingApi
import dev.zeekr.camerax_android.PigeonApiDynamicRangeProxyApi
import dev.zeekr.camerax_android.PigeonApiDynamicRangeUtilProxyApi

class DynamicRangeImpl(registrar: CameraXApiPigeonProxyApiRegistrar) : PigeonApiDynamicRangeProxyApi(registrar) {
    class UtilImpl(registrar: CameraXApiPigeonProxyApiRegistrar) : PigeonApiDynamicRangeUtilProxyApi(registrar) {
        override fun fromEncoding(value: Long): DynamicRangeEncodingApi {
            return Util.fromEncoding(value.toInt())
        }

        override fun fromBitDepth(value: Long): DynamicRangeBitDepthApi {
            return Util.fromBitDepth(value.toInt())
        }

        override fun toEncoding(api: DynamicRangeEncodingApi): Long {
            return Util.toEncoding(api).toLong()
        }

        override fun toBitDepth(api: DynamicRangeBitDepthApi): Long {
            return Util.toBitDepth(api).toLong()
        }
    }

    object Util {
        fun fromEncoding(value: Int): DynamicRangeEncodingApi {
            return when (value) {
                DynamicRange.ENCODING_UNSPECIFIED -> DynamicRangeEncodingApi.UNSPECIFIED
                DynamicRange.ENCODING_SDR -> DynamicRangeEncodingApi.SDR
                DynamicRange.ENCODING_HDR_UNSPECIFIED -> DynamicRangeEncodingApi.HDR_UNSPECIFIED
                DynamicRange.ENCODING_HLG -> DynamicRangeEncodingApi.HLG
                DynamicRange.ENCODING_HDR10 -> DynamicRangeEncodingApi.HDR10
                DynamicRange.ENCODING_HDR10_PLUS -> DynamicRangeEncodingApi.HDR10PLUS
                DynamicRange.ENCODING_DOLBY_VISION -> DynamicRangeEncodingApi.DOLBY_VISION
                else -> throw NotImplementedError("Not implemented value: $this")
            }
        }

        fun fromBitDepth(value: Int): DynamicRangeBitDepthApi {
            return when (value) {
                DynamicRange.BIT_DEPTH_UNSPECIFIED -> DynamicRangeBitDepthApi.UNSPECIFIED
                DynamicRange.BIT_DEPTH_8_BIT -> DynamicRangeBitDepthApi.EIGHT_BIT
                DynamicRange.BIT_DEPTH_10_BIT -> DynamicRangeBitDepthApi.TEN_BIT
                else -> throw NotImplementedError("Not implemented value: $this")
            }
        }

        fun toEncoding(api: DynamicRangeEncodingApi): Int {
            return when (api) {
                DynamicRangeEncodingApi.UNSPECIFIED -> DynamicRange.ENCODING_UNSPECIFIED
                DynamicRangeEncodingApi.SDR -> DynamicRange.ENCODING_SDR
                DynamicRangeEncodingApi.HDR_UNSPECIFIED -> DynamicRange.ENCODING_HDR_UNSPECIFIED
                DynamicRangeEncodingApi.HLG -> DynamicRange.ENCODING_HLG
                DynamicRangeEncodingApi.HDR10 -> DynamicRange.ENCODING_HDR10
                DynamicRangeEncodingApi.HDR10PLUS -> DynamicRange.ENCODING_HDR10_PLUS
                DynamicRangeEncodingApi.DOLBY_VISION -> DynamicRange.ENCODING_DOLBY_VISION
            }
        }

        fun toBitDepth(api: DynamicRangeBitDepthApi): Int {
            return when (api) {
                DynamicRangeBitDepthApi.UNSPECIFIED -> DynamicRange.BIT_DEPTH_UNSPECIFIED
                DynamicRangeBitDepthApi.EIGHT_BIT -> DynamicRange.BIT_DEPTH_8_BIT
                DynamicRangeBitDepthApi.TEN_BIT -> DynamicRange.BIT_DEPTH_10_BIT
            }
        }
    }

    override fun pigeon_defaultConstructor(encoding: Long, bitDepth: Long): DynamicRange {
        return DynamicRange(encoding.toInt(), bitDepth.toInt())
    }

    override fun unspecifid(): DynamicRange {
        return DynamicRange.UNSPECIFIED
    }

    override fun sdr(): DynamicRange {
        return DynamicRange.SDR
    }

    override fun hdrUnspecified10Bit(): DynamicRange {
        return DynamicRange.HDR_UNSPECIFIED_10_BIT
    }

    override fun hdr10_10Bit(): DynamicRange {
        return DynamicRange.HDR10_10_BIT
    }

    override fun hdr10Plus10Bit(): DynamicRange {
        return DynamicRange.HDR10_PLUS_10_BIT
    }

    override fun hlg10Bit(): DynamicRange {
        return DynamicRange.HLG_10_BIT
    }

    override fun dolbyVision8Bit(): DynamicRange {
        return DynamicRange.DOLBY_VISION_8_BIT
    }

    override fun dolbyVision10Bit(): DynamicRange {
        return DynamicRange.DOLBY_VISION_10_BIT
    }

    override fun getBitDepth(pigeon_instance: DynamicRange): Long {
        return pigeon_instance.bitDepth.toLong()
    }

    override fun getEncoding(pigeon_instance: DynamicRange): Long {
        return pigeon_instance.encoding.toLong()
    }
}
